### PR TITLE
Optimize toFuture

### PR DESCRIPTION
**Pull request by RaasAhsan**
This might be unnecessary, since the `Future` passed into `fromFuture` may or may not already be evaluated and depending on how fast it completes. So it would almost always go through the unoptimized path, unless the `Future` was constructed via `successful`.

The fiber one might be good because it keeps the future around to observe the result.

### CODE REVIEW

File: interop/shared/src/main/scala/scalaz/zio/interop/future.scala [hunk 1]
```diff
@@ -1,38 +1,65 @@
...
     def fromFuture[A](ftr: () => Future[A])(ec: ExecutionContext): IO[Throwable, A] =
-      IO.async { (cb: ExitResult[Throwable, A] => Unit) =>
-        ftr().onComplete {
-          case Success(a) => cb(ExitResult.succeeded(a))
-          case Failure(t) => cb(ExitResult.checked(t))
-        }(ec)
+      IO.suspend {
+        val f = ftr()
+        f.value match {
+          case Some(result) =>
```
Comment (hepin1989) [inline interop/shared/src/main/scala/scalaz/zio/interop/future.scala]: nice！

File: interop/shared/src/main/scala/scalaz/zio/interop/future.scala [hunk 2]
```diff
@@ -1,38 +1,64 @@
...
+            result match {
```
Comment (regiskuckaertz) [inline interop/shared/src/main/scala/scalaz/zio/interop/future.scala]: Using `fold` over pattern matching has become kind of the idiomatic style in ZIO, what do you think of
```scala
result.fold(t => IO.fail0(Cause.checked(t)), IO.now)
```

File: interop/shared/src/main/scala/scalaz/zio/interop/future.scala [hunk 3]
```diff
@@ -1,38 +1,64 @@
...
+                    case Failure(t) => cb(ExitResult.succeeded(ExitResult.checked(t)))
+                  }(ec)
+                }
+            }
           }
-        }
+
+        def poll: IO[Nothing, Option[ExitResult[Throwable, A]]] =
+          IO.sync {
+            ftr.value map {
```
Comment (regiskuckaertz) [inline interop/shared/src/main/scala/scalaz/zio/interop/future.scala]: Same here:
```scala
ftr.value.map(ExitResult.checked(_), ExitResult.succeeded(_))
```

Comment (regiskuckaertz) [reply interop/shared/src/main/scala/scalaz/zio/interop/future.scala]: ... or we could add `ExitResult#fromTry`

Comment (jdegoes) [reply interop/shared/src/main/scala/scalaz/zio/interop/future.scala]: Also: `fromEither`, `fromOption`.

Comment (RaasAhsan) [reply interop/shared/src/main/scala/scalaz/zio/interop/future.scala]: There isn't a clear way to implement `fromOption` -- what `Cause` do we map the `None` case to?

Comment (jdegoes) [reply interop/shared/src/main/scala/scalaz/zio/interop/future.scala]: `Cause[Unit]`.

**Comment by RaasAhsan** [comment]
Sorry about all the commits :D Haven't been able to run tests lately, SBT runs out of metaspace too fast now

**Comment by jdegoes** [comment]
Do you use the `sbt` script in the repo? It can reliably run tests.

**Comment by RaasAhsan** [comment]
didn't know about that one. thanks a bunch! works like a charm

**Comment by RaasAhsan** [comment]
scala 2.11 and linting on test scope -_- should scalafmt run on test scopes by default?

**Comment by jdegoes** [comment]
@RaasAhsan That would be nice. There's an alias you can use `fmt` that does both, but doesn't do `scalafmtSbt`.

**Comment by jdegoes** [comment]
@RaasAhsan Just a few conflicts and then we can get this one merged! Thanks for your work on it.

**Comment by jdegoes** [comment]
@RaasAhsan Thank you for the improvement!

**Pull request merged**